### ISSUE TITLE
[ADCS - Certificate] Can't connect on some hubs

### ISSUE THREAD

**Issue by Kcchouette** [airdcpp-web/airdcpp-webclient#333]
### Current UI and client versions:

AirDC++w 2.6.1 x86_64 (portable version)

### Operating system:

debian 10.1
OpenSSL 1.1.1c 

### Steps to reproduce the issue:

On a **new instance** of airdcpp-web (theses problems don't occur on my old one on debian 9).

Adding a hub using favorite hubs > Add new

When trying to connect to some hubs (not all) have theses error:

#### first kind of error

```
Connecting to adcs://hub.dcbase.org:16591 ...
TLS error: certificate has expired
```
How can I fix that? Only by asking the hub owner (and we can't speak the hub owner using airdcpp as we can't connect)?

#### second kind of error

```
Connecting to adcs://XXXXX:XXXX ...
TLS error: self signed certificate
```
That's true, it's a self-signed certificate. How can I go beyond that to connect even if it's a self-signed certificated?

#### third kind of error

Same hub than the *second kind of error*, but adding the kp at the end of the address
```
Connecting to adcs://XXXXX:XXXX?kp=SHA256/YYYYYYYYYYYYYYYYYYYYYYYYYYYY ...
TLS error: error:141A318A:SSL routines:tls_process_ske_dhe:dh key too small
```

... How can I go beyond that to connect on the hub?

**Comment by Kcchouette** [comment]
Version 2.5.1 and I think before does not have this bug

**Comment by maksis** [comment]
This might be related to the OpenSSL version of the portable version (your local OpenSSL version shouldn't matter as the portable version includes it's own libraries)

**Comment by maksis** [comment]
The application should also be working fine with self signed certificates. I'm going to need the hub addresses for issues 2 and 3 in order to be able to reproduce them.

**Comment by Kcchouette** [comment]
@maksis For the 2 and 3rd problem, best option is that I can test some snapshot version.
2nd option is I can ask for you a temporary account (I'm not the owner of the hub)

**Comment by maksis** [comment]
> 2nd option is I can ask for you a temporary account

I have no need for an account, the address is enough

**Comment by Kcchouette** [comment]
Ok, I'll paste it to you through your hub

**Comment by Kcchouette** [comment]
To sum-up test we made: I upgraded to `airdcpp_2.7.0b-19-g8245_webui-2.7.0-beta.2_64-beta2`, check that the two checkbox of bottom of connectivity > encryption are enable, and test if case 2 occurs → Sadly it occurs :(

**Comment by Kcchouette** [comment]
This issue seems very specific, it **only** occurs on debian (maybe only the version 10 and later or 10.1).

**Comment by maksis** [comment]
Seems like new security restrictions were added in Debian 10 for OpenSSL 1.1.1: https://wiki.debian.org/ContinuousIntegration/TriagingTips/openssl-1.1.1

Even though the static builds use their own OpenSSL library, they still seem to read the system-wide OpenSSL config file (`/etc/ssl/openssl.cnf`). For the author of the issue, the config value `CipherString = DEFAULT@SECLEVEL=2` caused the certificate to be rejected because the length of the DH was too small (https://weakdh.org). The solution is to comment out that config line, or even better, ask the hub owner to generate a new certificate for the hub with a longer DH key. 

Note that Debian 10 also blocks TLS versions older than 1.2 by default. That may also break encrypted file transfers between AirDC++ and outdated clients that only support TLS 1.0. Such restriction can be disabled by commenting out the line `MinProtocol = TLSv1.2` in `/etc/ssl/openssl.cnf`.

**Comment by Kcchouette** [comment]
To test a certificate:
` $ echo | openssl s_client -connect <address:port> -msg | grep "Verify return code"`

And check the reason

**Comment by Kcchouette** [comment]
Some feedback about the issue:
 - **first kind of error** →fixed by checking the two checkbox, need the 2.7.0 version or later of airdcpp-web; the owner of the hub @Ullner does not want to change the certificate for the moment
 - **second kind of error** and **third kind of error**: reporting the issue for the hub software itself and adding a patch that fixes this issue when renewing the certificate of the hub →https://bugs.launchpad.net/adchpp/+bug/1850053
